vscode-dotnet-runtime-library/src/Acquisition/LinuxVersionResolver.ts

```diff
@@ -0,0 +1,288 @@
+            {
+                const error = new DotnetAcquisitionDistroUnknownError(new Error(this.baseUnsupportedDistroErrorMessage));
+                this.acquisitionContext.eventStream.post(error);
+                throw error;
+            }
+
+            const pair : DistroVersionPair = { distro : distroName, version : distroVersion };
+            return pair;
+        }
+        catch(error)
+        {
+            const err = new DotnetAcquisitionDistroUnknownError(new Error(`${this.baseUnsupportedDistroErrorMessage} ... does /etc/os-release exist?`));
+            this.acquisitionContext.eventStream.post(err);
+            throw err;
+        }
+    }
+
+
+    /**
+     * @remarks we need to call the command executor to allow us to test this class.
+     * The command executor needs to be able to run async commands. We cant call async commands in the constructor.
+     * So we have to use this pattern of calling initialize to verify our members are not null in each function.
+     */
+    public async Initialize()
+    {
+        if(!this.distro)
+        {
+            this.distro = await this.getRunningDistro();
+        }
+
+        if(!this.distroSDKProvider)
+        {
+            this.distroSDKProvider = this.DistroProviderFactory(this.distro);
+        }
+
+        if(!this.distro || !this.distroSDKProvider)
+        {
+            const error = new DotnetAcquisitionDistroUnknownError(new Error(`${this.baseUnsupportedDistroErrorMessage} ... we cannot initialize.`));
+            this.acquisitionContext.eventStream.post(error);
+            throw error;
+        }
+    }
+
+    private DistroProviderFactory(distroAndVersion : DistroVersionPair) : IDistroDotnetSDKProvider
+    {
+        switch(distroAndVersion)
+        {
+            // Implement any custom logic for a Distro Class in a new DistroSDKProvider and add it to the factory here.
+            case null:
+                const error = new DotnetAcquisitionDistroUnknownError(new Error(this.baseUnsupportedDistroErrorMessage));
+                this.acquisitionContext.eventStream.post(error);
+                throw error;
+            default:
+                return new GenericDistroSDKProvider(distroAndVersion);
+        }
+    }
+
+    /**
+     *
+     * @param supportStatus The support status of this distro and version pair.
+     * @param fullySpecifiedDotnetVersion The version of dotnet requested to install, upgrade, etc.
+     * @remarks Throws a specific error below if a conflicting install type of dotnet exists on linux.
+     * Microsoft and distro feed packages together cause system instability with dotnet, so we dont want to let people get into those states.
+     * Eventually, we could add logic to remove them for users, but that may require consent first.
+     */
+    public async VerifyNoConflictInstallTypeExists(supportStatus : DotnetDistroSupportStatus, fullySpecifiedDotnetVersion : string) : Promise<void>
+    {
+        await this.Initialize();
```

Comment:
nit: initialize is called at the beginning of most public methods in this class and also before most of these methods are called in the LinuxGlobalInstaller. Not a big deal since the method will be so fast on subsequent calls, but I would choose a pattern- either the caller is expected to initialize before calling anything else, or the class does it internally and the caller doesn't have to care (in which case it could be private)

Reply:
I don't really like what I was able to come up with to resolve this, but I ended up making it so initialize is called privately only and you need to call a method to access the distro functions from within this class. I dont like the pattern of relying on the caller to know to call initialize as it requires you to know how the internals of the code work

Reply:
Although, this pattern is likely to be slow if these calls end up being no-ops.

Reply:
Ya, mostly they will be. I don't like it either. I think a singleton pattern may help it a bit. I read up when doing this though and most people suggested to instead call an initialize method like so. This is in the particular case of needing an object to be asynchronously constructed (async constructors are not allowed)

There is probably a more elegant way to do it, but I am not too invested in cleaning this up right now